js/treetest-edit.js

```diff
@@ -20,24 +20,59 @@ function bindFunctions(){
     	$('#form').submit()
 	});
 
-  $("#createNewRootNode").click(function() {
+  	$("#createNewRootNode").click(function() {
   		$('#tree').jstree().create_node("#", "New Root Item", "last");
 	});
 
 	//expand parent node when child is created
-  $("#createNewNode").click(function() {
-  	var selectedParent = $("#tree").jstree("get_selected");
-  	$('#tree').jstree().create_node(selectedParent, "New Item", "last");
-  	$("#tree").jstree("open_node", selectedParent);
+  	$("#createNewNode").click(function() {
+  		var selectedParent = $("#tree").jstree("get_selected");
+  		$('#tree').jstree().create_node(selectedParent, "New Item", "last");
+  		$("#tree").jstree("open_node", selectedParent);
 	});
 
-  $("#expandAll").click(function() {
-  	$("#tree").jstree('open_all');
+  	$("#expandAll").click(function() {
+  		$("#tree").jstree('open_all');
 	});
 
 	$('#tree').on("changed.jstree", function (e, data) {
 		$("#tree").jstree("toggle_node", data.selected);
 	});
+
+	$("#newTask").on("click", function() {
+		var newRow = document.getElementById('newTask').querySelector('template').content.cloneNode(true);
+		$("#tasks").children("tbody").append($(newRow));
+	});
+
+	$("#tasks").on("click", ".deleteTaskBtn", function () {
+        $(this).closest("tr").remove();   
+    });
+
+	$("#tasks").on("click", ".addAnswerBtn", function () {
+		var node = $("#tree").jstree("get_selected");
+		var nodes = $('#tree').jstree('get_path', node, '/').split('/');
+
+		expectedAnswersElement=$(this).closest('td').find('.expectedAnswers');
+		expectedAnswers = JSON.parse(expectedAnswersElement.val());
+		expectedAnswers.push(nodes);
+		expectedAnswersElement.val(JSON.stringify(expectedAnswers));	
+
+		var answerElement = document.getElementById('newAnswer').querySelector('template').content.cloneNode(true);
+		answerElement.querySelector('p').textContent=JSON.stringify(nodes);
+		$(this).closest('tr').find('.list-group').append($(answerElement));	
+    });
+
+	$("#tasks").on("click", ".deleteAnswerBtn", function () {
+		const expectedAnswerContainer = $(this).closest("div");
+		const divIndex = expectedAnswerContainer.index();
+
+		const expectedAnswersElement=$(this).closest('td').find('.expectedAnswers');
+		expectedAnswers = JSON.parse(expectedAnswersElement.val());		
+		expectedAnswers.splice(divIndex, 1);
+
+		expectedAnswersElement.val(JSON.stringify(expectedAnswers));
+		expectedAnswerContainer.remove();

```

Comment:
If I understand this right, the algorithm is:

- Find out what the index of the current list item in the `.list-group` is: call that _n_
- Remove the _n_<sup>th</sup> item in the list of expected answers in the `.expectedAnswers` input

This works as expected, as long as the order of the expected answers in the `.expectedAnswers` input always stays the same as the list in the `.list-group`, _ie_ the _n_<sup>th</sup> list item in the `.list-group` always corresponds to the _n_<sup>th</sup> list item in the value of the `.expectedAnswers` input.

As far as I can tell, that's a reasonable assumption, given the way the code works right now: we always add items in order, and always remove items in order. If the code works as expected, and the user doesn't meddle with it, it should work.

Still, it could be more robust if instead the algorithm was: remove the first entry in the list in the `.expectedAnswers` input which matches the content of the current list item in the `.list-group`. Then you wouldn't have to assume that the list ordering has been maintained

Reply:
Yeah, I did try various approaches and this one seems to be the best considering that we are ensuring the order is maintained. The approach you suggest will have problems if we have duplicate entries.

Let's say a question has following expected answers:
1. ["Vegetables", "Tomato"]
2. ["Fruits", "Tomato"]
3. ["Vegetables", "Tomato"]

User realizes that it has duplicate entries and he removes the last item from the list. On the UI it now shows the following:
1. ["Vegetables", "Tomato"]
2. ["Fruits", "Tomato"]

However the hidden field for storing expected answers have the data saved as - [["Fruits", "Tomato"], ["Vegetables", "Tomato"]]. If the user now saves the data and comes back to edit the answers they will be displayed in the reverse order.

Reply:
This is not a big deal since it will not cause any issues, but it somehow feels odd to me

Reply:
I think it makes sense: the hidden list and the display list contain all the same entries, and we'd have no need to guarantee that they are in the same order